### PR TITLE
CLOUDP-97418: Creates one serverless instance in the specified project.

### PR DESCRIPTION
**Pull request by andreaangiolillo**
## Proposed changes

_Jira ticket:_ CLOUDP-97418

## Checklist

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation (if appropriate)
- [x] I have updated [e2e/E2E-TESTS.md](e2e/E2E-TESTS.md) (if a new command or e2e test has been added)
- [x] I have run `make fmt` and formatted my code

## Further comments

Evergreen Task: [atlas_serverless_e2e[E2E Atlas Cluster Tests]](https://evergreen.mongodb.com/task/mongocli_master_e2e_atlas_clusters_atlas_serverless_e2e_patch_93e5f9eef735dac643b7205e7a0e6be9b47e162e_61150ecf850e611f446793f2_21_08_12_12_06_41)

```
./bin/mongocli atlas sc create test2 --provider AWS --region US_EAST_1
Serverless Cluster test2 created.
```
```
./bin/mongocli atlas sc create test --provider AWS --region US_EAST_1 -o json
{
  "id": "6114f8f9461a6010cad4e1d2",
  "groupId": "5e4e593f70dfbf1010295836",
  "mongoDBVersion": "5.0.2",
  "name": "test",
  "createDate": "2021-08-12T10:33:29Z",
  "providerSettings": {
    "backingProviderName": "AWS",
    "providerName": "SERVERLESS",
    "regionName": "US_EAST_1"
  },
  "stateName": "CREATING",
  "connectionStrings": {},
  "links": [
    {
      "rel": "self",
      "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/groups/5e4e593f70dfbf1010295836/serverless/test"
    },
    {
      "rel": "http://cloud.mongodb.com/restoreJobs",
      "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/groups/5e4e593f70dfbf1010295836/serverless/test/restoreJobs"
    },
    {
      "rel": "http://cloud.mongodb.com/snapshots",
      "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/groups/5e4e593f70dfbf1010295836/serverless/test/snapshots"
    }
  ]
}
```

### CODE REVIEW

File: internal/cli/atlas/serverlessclusters/watch.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+		Use:   "watch <clusterName>",
+		Short: "Watch for a serverless cluster to be available.",
+		Args:  require.ExactArgs(1),
+		Annotations: map[string]string{
+			"args":            "clusterName",
+			"clusterNameDesc": "Name of the cluster to watch.",
+		},
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			return opts.PreRunE(
+				opts.ValidateProjectID,
```
Comment (colm-quinn) [inline internal/cli/atlas/serverlessclusters/watch.go:69]: [q] Is `ValidateProjectID` correct? (all other references in the file are to `projectId`

File: internal/cli/atlas/serverlessclusters/create.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+		ProviderSettings: &atlas.ServerlessProviderSettings{
+			BackingProviderName: opts.provider,
+			ProviderName:        providerName,
+			RegionName:          opts.region,
+		},
+	}
+}
+
+// mongocli atlas serverlessCluster(s)|sc create <clusterName> --backingProviderName backingProviderName --providerName providerName --regionName regionName [--projectId projectId].
+func CreateBuilder() *cobra.Command {
```
Comment (colm-quinn) [inline internal/cli/atlas/serverlessclusters/create.go:68]: [late to the party q] How do we ensure these commands only exist on Atlas and don't appear on the OM/CM cli?

Comment (threebee) [reply internal/cli/atlas/serverlessclusters/create.go:68]: Afaik, there are separate directories in `internal/cli` directory. `atlas, opasmanager, cloudmanager`. Each of them has a set of sub commands which are available under them:

Example for `cloudmanager`:
```
	cmd.AddCommand(
		clusters.Builder(),
		alerts.Builder(),
		backup.Builder(),
		servers.Builder(),
		automation.Builder(),
		security.Builder(),
		dbusers.Builder(),
		events.Builder(),
		monitoring.Builder(),
		processes.Builder(),
		metrics.Builder(),
		logs.Builder(),
		agents.Builder(),
		maintenance.Builder(),
		performanceadvisor.Builder(),
		featurepolicies.Builder(),
```

File: internal/cli/atlas/serverlessclusters/watch.go (hunk 1 above)
Comment (andreaangiolillo) [reply internal/cli/atlas/serverlessclusters/watch.go:69]: yes, it is. We validate the project id inside the `opts.PreRunE` also in the other command.
`ValidateProjectID` checks that one the user has provided the `--projectId` flag or that the used profile has the `projectId`

```golang
// ValidateProjectID validates projectID.
func (opts *GlobalOpts) ValidateProjectID() error {
	if opts.ConfigProjectID() == "" {
		return errMissingProjectID
	}
	return validate.ObjectID(opts.ConfigProjectID())
}

// ConfigProjectID returns the project id.
// If the id is empty, it caches it after querying config.
func (opts *GlobalOpts) ConfigProjectID() string {
	if opts.ProjectID != "" {
		return opts.ProjectID
	}
	opts.ProjectID = config.ProjectID()
	return opts.ProjectID
}
```

Comment (threebee) [reply internal/cli/atlas/serverlessclusters/watch.go:69]: I _think_ it is correct. It is defined in https://github.com/mongodb/mongocli/blob/fcf9f4ee31467445cdbe80cbdca1392444ea2bde/internal/cli/global_opts.go#L66

And all the methods there end with `ProjectID`. Maybe it is just a styling issue.

**Review by colm-quinn — APPROVED** [review]
LGTM, thanks for the answers!

**Review by threebee — APPROVED** [review]
LGTM!

File: e2e/atlas/serverless_clusters_test.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+			atlasEntity,
+			serverlessClustersEntity,
+			"create",
+			clusterName,
+			"--region=US_EAST_1",
+			"--provider=AWS",
+			"-o=json")
+		cmd.Env = os.Environ()
+		resp, err := cmd.CombinedOutput()
+		req.NoError(err)
```
Comment (gssbzn) [inline e2e/atlas/serverless_clusters_test.go:50]: @fmenezes while you fix the tests on your PR can you fix this line, we are losing the command output which is extremely helpful, change this to `req.NoError(err, string(resp))` to get the command output, same for other commands that disregard it

**Pull request merged**